From 89d94b6b98b96674ca1ac13e7bb7e822ef1edb48 Mon Sep 17 00:00:00 2001
From: Yang Yingliang <yangyingliang@huawei.com>
Date: Thu, 4 Feb 2021 09:48:57 +0800
Subject: [PATCH] vdso: do cntvct workaround in the VDSO
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 89d94b6b98b96674ca1ac13e7bb7e822ef1edb48


hulk inclusion
category: feature
bugzilla: 47984
CVE: NA

--------------------------------------------------

If a cntvct workaround is enabled, read CNTVCT_EL0 twice
in VDSO to avoid the clock bug.

Without this patch on Kunpeng916:
./gettimeofday -E -C 200 -L -S -W -N "gettimeofday"
Running:        gettimeofday# ./gettimeofday -E -C 200 -L -S -W -N gettimeofday
             prc thr   usecs/call      samples   errors cnt/samp
gettimeofday   1   1      0.31753          198        0    20000

With this patch on Kunpeng916:
./gettimeofday -E -C 200 -L -S -W -N "gettimeofday"
Running:        gettimeofday# ./gettimeofday -E -C 200 -L -S -W -N gettimeofday
             prc thr   usecs/call      samples   errors cnt/samp
gettimeofday   1   1      0.05244          198        0    20000

Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/include/asm/vdso/gettimeofday.h | 11 +++++++++++
 1 file changed, 11 insertions(+)

--- a/arch/arm64/include/asm/vdso/gettimeofday.h
+++ b/arch/arm64/include/asm/vdso/gettimeofday.h
@@ -83,6 +83,17 @@ static __always_inline u64 __arch_get_hw
 	 */
 	isb();
 	asm volatile("mrs %0, cntvct_el0" : "=r" (res) :: "memory");
+	if (vd->vdso_fix) {
+		u64 new;
+		int retries = 50;
+
+		asm volatile("mrs %0, cntvct_el0" : "=r" (new) :: "memory");
+		while (unlikely((new - res) >> vd->vdso_shift) && retries) {
+			asm volatile("mrs %0, cntvct_el0" : "=r" (res) :: "memory");
+			asm volatile("mrs %0, cntvct_el0" : "=r" (new) :: "memory");
+			retries--;
+		}
+	}
 	arch_counter_enforce_ordering(res);
 
 	return res;
